protocols/dcutr/examples/client.rs

```diff
@@ -197,25 +197,41 @@ fn main() -> Result<(), Box<dyn Error>> {
         }
     }
 
-    // Wait till connected to relay to learn external address.
+    // Wait till connected to relay to learn external address. In case we are in listening mode,
+    // wait for the relay to accept our reservation request.
     block_on(async {
+        let mut learned_observed_addr = false;
+        let mut relay_accepted_reservation = false;
+
         loop {
             match swarm.next().await.unwrap() {
                 SwarmEvent::NewListenAddr { .. } => {}
                 SwarmEvent::Dialing { .. } => {}
                 SwarmEvent::ConnectionEstablished { .. } => {}
                 SwarmEvent::Behaviour(Event::Ping(_)) => {}
+                SwarmEvent::Behaviour(Event::Relay(client::Event::ReservationReqAccepted {
+                    ..
+                })) => {
+                    info!("Relay accepted our reservation request.");
+                    relay_accepted_reservation = true
+                }
                 SwarmEvent::Behaviour(Event::Relay(_)) => {}
                 SwarmEvent::Behaviour(Event::Identify(IdentifyEvent::Sent { .. })) => {}
                 SwarmEvent::Behaviour(Event::Identify(IdentifyEvent::Received {
                     info: IdentifyInfo { observed_addr, .. },
                     ..
                 })) => {
-                    info!("Observed address: {:?}", observed_addr);
-                    break;
+                    info!("Relay observes us under the address: {:?}", observed_addr);
+                    learned_observed_addr = true;
                 }
                 event => panic!("{:?}", event),
             }
+
+            if learned_observed_addr
+                && (matches!(opts.mode, Mode::Dial) || relay_accepted_reservation)
```

Comment:
Would this be easier to understand if we matched explicitly on `Mode` and returned `true` for dial and `relay_accepted_reservation` for listen?

Reply:
What do you think of https://github.com/libp2p/rust-libp2p/pull/2642/commits/75697c253d774294abdf2a00a40718f8a6d32493?

Reply:
Much better, thank you!